vm/src/obj/objstr.rs

```diff
@@ -48,7 +48,39 @@ impl TryIntoRef<PyString> for &str {
     }
 }
 
+#[pyclass(__inside_vm, name = "str")]
```

Comment:
It kinda bothers me that this (and the doc comment) is on the impl block instead of the struct - is there any way to make that work?

Reply:
Not the `#[pyclass]` attribute for methods, but if we made another trait `PyDocItem` or something then we could put the doc comment on the struct definition. I'd say it's probably not worth it.

Reply:
For the impl block, PyO3 uses `#[pymethods]` and `#[pyclass]` goes on the struct. I think we should imitate that design.

Reply:
For the impl block, PyO3 uses #[pymethods] and #[pyclass] goes on the struct. I think we should imitate that design.

Reply:
Actually I think maybe `#[pyimpl]` would be better? (since the impl block can contain more than just methods)

Reply:
Something like:
```rust
#[pyclass]
/// doc
struct PyString {
    value: String,
}
type PyStringRef = PyRef<PyString>;
#[pymethods]
impl PyStringRef {
    // ...
}
```
The problem is that the type referenced in the `impl` block is different from the type in the `struct` declaration, so it won't be able to associate easily between those two proc macros.

Reply:
As a short-term solution, maybe `pymethods` can also take the name of the class?

(When arbitrary-self-types is available, we can go back to `impl PyString` and clean it up.)

Reply:
I think for now I'll just keep it as is, because I don't think there's a lot of incentive to split it up between the two for not much return.

Reply:
I really don't think we should merge this without moving the `#[pyclass]` annotation to the type / struct.

- The type is what you are declaring as representing a python class, not the impl block.
- It's inconsistent with similar projects in the ecosystem.
- If we are going to allow `#[pymethod]` on methods in trait impls as well, which one would get the `#[pyclass]`?

In general, I think we need to take more time to work out the kinks for impactful changes like this to avoid unnecessary churn.

Reply:
> In general, I think we need to take more time to work out the kinks for impactful changes like this to avoid unnecessary churn.

I agree with that, but ~~it's really an implementation problem with doing separate `#[pyclass]` and `#[pyimpl]`. If we transform:~~

```rust
#[pyclass]
/// docstring
struct PyString {
    value: String,
}
```
to
```rust
struct PyString {
    value: String,
}
impl PyClassTrait for PyString {
    const DOC: &'static str = "docstring";
}
```
~~Then there's no way to access that implementation for PyString when `#[pyimpl]` is on a `impl PyStringRef` block - Py. We have three options here:~~
* ~~Keep as is, with the~~

Nevermind, we could `impl<T> PyClassTrait for PyRef<T> where T: PyClassTrait { const DOC: &'static str = T::DOC; }`. I'll do that.

Reply:
Not quite sure on what to call "`PyClassTrait`" either, do you have any ideas?

Reply:
I would prefer to keep rust doc strings for rust, and add python docstrings explicitly, like this:
```rust
#[pymethod(name="add", doc="Adds two things (this is a python docstring)")
// This is rust documentation
fn add(a: PyObjectRef, b: PyObjectRef) -> PyResult {
}
```

Reply:
The thing is, the functions probably aren't going to need Rust documentation, as they're just implementation details for the final Python API.

Reply:
Yes, I agree that this is the case. Still I think it is better to explicitly state that the documentation is for the python side.

Reply:
@windelbouwman I disagree.

- Putting the docs in a `#[pyclass(doc = "...")]` is pretty awkward to me because you have to manually specify the newlines. Also multi-line attributes are fairly uncommon and look out of place.
- We can use regular `//` comments for implementation-level comments.
- Using `///` comments also means that developers of native extensions can see the python docs from rustdoc.
- It has precedent in the rust community - many binding generators do the same thing.